akka-cluster/src/main/scala/akka/cluster/CrossDcClusterHeartbeat.scala

```diff
@@ -0,0 +1,365 @@
+      if (verboseHeartbeat) log.debug(s"$incLocalClockJson Cluster Node [{}][{}] - Trigger extra expected (cross) heartbeat from [{}]", selfAddress, from.address)
+      crossDcFailureDetector.heartbeat(from.address)
+    }
+
+  private def selfIsResponsibleForCrossDcHeartbeat(): Boolean =
+    {
+      val activeDcs: Int = dataCentersState.dataCenters.size
+      val moreThanOneDcActive = activeDcs > 1
+
+      val res = cluster.state.members
+        .find(_.uniqueAddress == selfUniqueAddress)
+      res.exists(_.status == MemberStatus.Up)
+
+      val shouldBeActive = dataCentersState.shouldActivelyMonitorNodes(selfDataCenter, selfUniqueAddress, s ⇒ log.info(s"$incLocalClockJson $s"))
+      moreThanOneDcActive && shouldBeActive
+    }
+
+  /** Idempotent, become active if this node is n-th oldest and should monitor other nodes */
+  private def becomeActiveIfResponsibleForHeartbeat(): Unit = {
+    if (!activelyMonitoring && selfIsResponsibleForCrossDcHeartbeat()) {
+      log.info(s"$incLocalClockJson Becoming ACTIVE in {}, monitoring nodes: {}", selfDataCenter, dataCentersState)
+      activelyMonitoring = true
+
+      context.become(active orElse introspecting)
+    } else if (!activelyMonitoring)
+      log.info(s"$incLocalClockJson Remaining DORMANT; others in {} handle monitoring for me ({})", selfDataCenter, dataCentersState.state.getOrElse(cluster.selfDataCenter, Set.empty).map(m ⇒ m.address.port.get + "@" + m.status).mkString(","))
+  }
+
+  /** Idempotent, become or remain dormant if this node does not have to fulful the role of actively monitoring */
+  private def becomeDormantIfPossible(): Unit = {
+    if (!selfIsResponsibleForCrossDcHeartbeat()) {
+      log.info(s"$incLocalClockJson Becoming DORMANT in {}, all known nodes: {}", selfDataCenter)
+      activelyMonitoring = false
+      context.become(dormant orElse introspecting)
+    }
+  }
+
+}
+
+/** INTERNAL API */
+@InternalApi
+private[akka] object CrossDcHeartbeatSender {
+
+  final case class ReportStatus()
+
+  sealed trait StatusReport
+  final case object RunningWithSingleDcThusNotMonitoringAcrossDcs extends StatusReport
+  final case class StartMonitoring(monitoredNodes: immutable.Set[UniqueAddress]) extends StatusReport
+
+  sealed trait MonitoringStateReport extends StatusReport
+  final case class MonitoringActive(state: CrossDcHeartbeatingState) extends MonitoringStateReport
+  final case class MonitoringDormant() extends MonitoringStateReport
+}
+
+/** INTERNAL API */
+@InternalApi
+private[cluster] final case class CrossDcHeartbeatingState(
+  failureDetector:         FailureDetectorRegistry[Address],
+  nrOfMonitoredNodesPerDc: Int,
+  state:                   Map[ClusterSettings.DataCenter, immutable.SortedSet[Member]]
+) {
+  import CrossDcHeartbeatingState._
+
+  // TODO should we optimise this to know about DC of it?
+  def contains(address: UniqueAddress): Boolean =
```

Comment:
isn't this only used for checking those that we are actively monitoring?
doesn't that correspond to `activeReceivers`?
perhaps make `activeReceivers` a val or lazy val (and a Set[UniqueAddress]) then this is solved